File: lab2/gr9sub1/Понтаплёв_Артём/src/fork_example.c
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <sys/wait.h>

#define CHILD_TIME 5
#define PARENT_TIME 7
int main()
{
printf("parent: start with PID=%d\n", getpid());
fflush(stdout);
for (int i = 0; i < 2; i++)
{
pid_t pid = fork();
if (pid == 0)
{
printf("child[%d]: PID=%d, PPID=%d\n", i, getpid(), getppid());
fflush(stdout);
sleep(CHILD_TIME);
exit(0);
}
else if (pid < 0)
{
perror("fork failed");
exit(1);
}
}
sleep(PARENT_TIME);
for (int i = 0; i < 2; i++)
{
int status;
pid_t child_pid = wait(&status);
if (child_pid > 0)
{
printf("parent: child %d exited with status %d\n", child_pid, WEXITSTATUS(status));
fflush(stdout);
}
}
printf("parent: all children completed\n");
fflush(stdout);
return 0;
}

File: lab2/gr9sub2/КАЛЕНИК_ИЛЬЯ/src/fork_example.py
import os
import sys
import time

def main():
    parent_pid = os.getpid()
    print(f"parent(PID={parent_pid})", flush=True)

    children = []
    labels = ['A', 'B']

    for i in range(2):
        pid = os.fork()
        if pid == 0:
            
            child_label = labels[i]
            my_pid = os.getpid()
            parent_pid = os.getppid()
            print(f"child_{child_label}: PID={my_pid}, PPID={parent_pid}", flush=True)

           
            time.sleep(15)

            print(f"child_{child_label}: exiting now", flush=True)
            sys.exit(i + 10)  

        else:
            
            children.append(pid)

    
    for child_pid in children:
        pid, status = os.waitpid(child_pid, 0)
        if hasattr(os, 'waitstatus_to_exitcode'):
            exit_code = os.waitstatus_to_exitcode(status)
        else:
            exit_code = status >> 8  

        print(f"parent: child PID={pid} exited with code {exit_code}", flush=True)

    print("parent: all children finished", flush=True)

if __name__ == "__main__":
    main()
